{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code to compare lineage-specific responsiveness [after treatment with external stressor (osteogenic differentiation medium)] in high D cells (FLAT) with low D cells (PILLAR) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Written by Ranya K. A. Virk & Vasundhara Agrawal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import all libraries\n",
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os,sys,glob\n",
    "import pandas as pd\n",
    "import time\n",
    "from sympy.solvers import solve\n",
    "from sympy import Symbol\n",
    "import sympy as sp\n",
    "import scipy as sc\n",
    "from scipy import stats\n",
    "from scipy.stats import binned_statistic\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from scipy.stats import norm\n",
    "from matplotlib.ticker import FormatStrFormatter\n",
    "import scipy.optimize \n",
    "import seaborn as sns\n",
    "from scipy.interpolate import make_interp_spline, BSpline\n",
    "from mpl_toolkits import mplot3d\n",
    "from matplotlib import rcParams\n",
    "from IPython.display import display\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Function Calculating Sensitivity (CPMC Model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################################################\n",
    "#### Calculate the change of expression due to the change of chromatin packing (CPMC)###\n",
    "# This Code has been previously been made available on GitHub: https://github.com/BiophotonicsNU/macrogenomics.\n",
    "##################################################################################\n",
    "\n",
    "def Se(D_fit):\n",
    "    '''\n",
    "    D_fit is the chromatin packing scaling\n",
    "    '''\n",
    "  \n",
    "    \n",
    "    ## Load MC simulation result\n",
    "    sec          =np.genfromtxt('second_derivative_TF_norm.csv')\n",
    "    mRNA_initial =np.genfromtxt('max_mRNA_initial.csv')\n",
    "    phi_initial  =np.genfromtxt('phi_initial.csv')\n",
    "    tot_con      =np.genfromtxt('tot_con.csv')\n",
    "    Li0          =15e-9 # Size of interaction volume for single base pair\n",
    "    r_min        =1e-9 # size of elementary particle\n",
    "    len_gene     =6e3 # average length of genes\n",
    "    initial_aver =0.00056 # the average initial expression rate\n",
    "    M_f = 1e6\n",
    "    Li           =(Li0+len_gene**(1/D_fit)*r_min) # the total interaction volume\n",
    "\n",
    "    ## Load the equation of Se\n",
    "    ratio = M_f#(r_max/r_min)**D_fit # M_f/M_min\n",
    "    sigma = phi_initial*(1-phi_initial) # variance of crowding density at each point\n",
    "    sigma_phi = sigma*(Li/r_min)**(D_fit-3) # variance of average crowding density in each interaction volume \n",
    "    d_sigma_phi = sigma_phi*(np.log(Li/r_min)+(3-D_fit)/D_fit**2*len_gene**(1/D_fit)*np.log(len_gene)*(r_min/Li)) # the derivative of sigma_phi as respect to D\n",
    "    Se = (1/D_fit**2*np.log(ratio)+1/2*sec/(1+1/2*sec*sigma_phi)*d_sigma_phi)*D_fit # gene transcription Sensitivity\n",
    "    initial_expression = initial_aver \n",
    "    expression = mRNA_initial*(1+1/2*sec*sigma_phi) # the average expression of gene share similar molecular properties\n",
    "    ## Fitting the model with poly10\n",
    "    x = np.log(expression/initial_aver)\n",
    "    y = Se\n",
    "    index_start=np.arange(x.shape[0])[~np.isnan(x)][0]\n",
    "    para2 = np.poly1d(np.polyfit(x[index_start:],y[index_start:],10))\n",
    "    x = np.linspace(-5,3,100) ## (-3.8,3) for 3D plot. (-5,3) for cross-section\n",
    "    return (np.exp(x)*initial_aver,para2(x))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All DE Genes in Induced MSCs compared to Control MSCs (p-adj<0.05)\n",
    "### Evaluating Experimental Results and CPMC Model Predictions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize = (15,11))\n",
    "\n",
    "##################################################################################\n",
    "#### Experimental Calculation of Lineage Specific Responsiveness Coefficient #####\n",
    "##################################################################################\n",
    "\n",
    "#Treatement groups: SF = MSCs on Flat; SP = MSCs on Pillars; IF = Induced MSCs on Flat; IP = Induced MSCs on Pillars\n",
    "#Load data (Differnetially expressed genes with average gene expression for each group, p<0.05):\n",
    "ALL_reg_genes = pd.read_csv('DE_Meanvalues.csv')\n",
    "ALL_mean_genes = pd.DataFrame(index = ALL_reg_genes.index, columns = ['IF','IP','SF','SP'])\n",
    "ALL_mean_genes.IF = ALL_reg_genes ['IF']\n",
    "ALL_mean_genes.SF = ALL_reg_genes ['SF']\n",
    "ALL_mean_genes.IP = ALL_reg_genes ['IP']\n",
    "ALL_mean_genes.SP = ALL_reg_genes ['SP']\n",
    "\n",
    "#Expression Relative to control (uninduced MSCs on Flat):\n",
    "ALL_expression_norm=(np.log2(ALL_mean_genes.div(ALL_mean_genes['SF'],axis=0)))\n",
    "\n",
    "## Divide control into 2 quantile subgroups based on initial expression of stem state:\n",
    "lb=['Experiment\\nLow Initial Expression','Experiment\\nHigh Initial Expression']\n",
    "quant_con = np.percentile(ALL_mean_genes['SF'],[12,50,60,98])\n",
    "color_index=[1,4]\n",
    "for i in range(2):\n",
    "    \n",
    "    display(i)\n",
    "    gene_con = (ALL_mean_genes['SF']<quant_con[i*2+1]) & (ALL_mean_genes['SF']>quant_con[i*2])\n",
    "    expression_con_norm= ALL_expression_norm[gene_con]\n",
    "    expression_con=ALL_mean_genes[gene_con]\n",
    "    \n",
    "    ## divide induced MSCs into decile subgroups\n",
    "    quant_inducedMSCs = np.percentile(expression_con_norm['IF'],np.linspace(0,100,11))\n",
    "    y = []\n",
    "    error=[]\n",
    "\n",
    "    for j in range(10):\n",
    "        gene = (expression_con_norm['IF']<quant_inducedMSCs[j+1]) & (expression_con_norm['IF']>quant_inducedMSCs[j])\n",
    "        de = expression_con['IF'][gene]/expression_con['SF'][gene]\n",
    "        nu = expression_con['IP'][gene]/expression_con['SP'][gene]\n",
    "        cov=np.cov(nu,de)[1,0] \n",
    "        \n",
    "        ##### Evaluate lineage-specific reponsiveness as the ratio of expression rate of (IP/SP)/(IF/SF):\n",
    "        y.append(np.mean(nu)/np.mean(de))\n",
    "        error.append((np.mean(nu)/np.mean(de))*(np.var(nu)/np.mean(nu)**2+np.var(de)/np.mean(de)**2\\\n",
    "                     -2*cov/np.mean(de)/np.mean(nu))**(1/2)/np.size(nu)**(1/2))\n",
    "    ax.errorbar(quant_inducedMSCs[1:],y,markersize=25,yerr=error,fmt='.',capsize=7,label=lb[i],color=sns.color_palette(\"muted\")[color_index[i]])\n",
    "    ax.set_ylabel(r'$R_{LS}$',fontsize=38)\n",
    "    ax.set_xlabel(r'$log_2(E_{IF}/E_{SF})$',fontsize=38,weight='bold')\n",
    "\n",
    "##################################################################################\n",
    "##### CPMC Model Prediction of Lineage Specific Responsiveness Coefficient #######\n",
    "##################################################################################\n",
    "\n",
    "#Range of D measured in MSCs:\n",
    "D_0 = 2.85                        \n",
    "D_i = np.linspace(D_0,2.58,100)  \n",
    "dD =D_i[1]-D_i[0]\n",
    "\n",
    "# Calculate CPMC predicted sensitivity (dlnE/dlnD) for each D value in given range (3D array)\n",
    "Se_all=[]\n",
    "for i in range(100):\n",
    "    temp,Se_temp=Se(D_i[i])\n",
    "    Se_all.append(Se_temp)\n",
    "inte=np.exp((((np.array(Se_all).transpose()/D_i).transpose())*dD).sum(0)) \n",
    "\n",
    "lb=['CPMC Model\\nLow Initial Expression','CPMC Model\\nHigh Initial Expression']\n",
    "color_index=[1,4]\n",
    "x,temp = Se(D_0)\n",
    "y,temp = Se(D_0)\n",
    "ind=[70,85] # Indices representing the relative initial expression of the stem flat state for initially lowly expressed\n",
    "            # and initially highly expressed genes, used to subset 2D slices from 3D sensitivity curve\n",
    "xlim=[-2,5,-2,5]  \n",
    "for j,i in enumerate(ind):\n",
    "    # Calculate predicted expression rate for lower D chromatin\n",
    "    b_1_0 = x[i] \n",
    "    b_2_0 = y\n",
    "    db_0= b_2_0/b_1_0\n",
    "\n",
    "    # Calculate predicted expression for higher D chromatin\n",
    "    x_high_D=x[i]*inte[i]\n",
    "    y_high_D= y*inte\n",
    "    db=y_high_D/x_high_D\n",
    "    \n",
    "    # Calculate predicted lineage specific responsiveness coefficient \n",
    "    b_rate = db_0/db\n",
    "    \n",
    "    # Plot CPMC predictions within a given range of np.log2(b_2_0/b_1_0)\n",
    "    x_plot=np.log2(b_2_0/b_1_0)\n",
    "    y_plot=1/b_rate\n",
    "    ind=(x_plot>xlim[j*2])&(x_plot<=xlim[j*2+1])\n",
    "    x_plot=x_plot[ind]\n",
    "    y_plot=y_plot[ind]\n",
    "    ax.plot(x_plot, y_plot,label=lb[j],c=sns.color_palette('muted')[color_index[j]])\n",
    "\n",
    "ax.legend(frameon=True,fontsize=20)\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.set_xlim(-2.5,6)\n",
    "ax.set_ylim(0.5,1.6)\n",
    "plt.autoscale(False)\n",
    "ax.grid(False)\n",
    "for axis in ['bottom','right']:\n",
    "    ax.spines[axis].set_linewidth(3)\n",
    "    ax.spines[axis].set_color('k')\n",
    "\n",
    "ax.set_facecolor('white')\n",
    "ax.yaxis.set_ticks_position('left')\n",
    "ax.xaxis.set_ticks_position('bottom')\n",
    "plt.xticks(fontsize=29,weight='bold')\n",
    "plt.yticks(fontsize=29,weight='bold')\n",
    "fig.savefig('ModelA-2.85-2.58_70-85_'+str(ind[0])+'_'+str(ind[1])+'.tif',dpi=1000)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
